libbeat/processors/dissect/field.go

```diff
@@ -0,0 +1,206 @@
+	return true
+}
+
+func (f baseField) String() string {
+	return fmt.Sprintf("field: %s, ordinal: %d, greedy: %v", f.key, f.ordinal, f.IsGreedy())
+}
+
+type normalField struct {
+	baseField
+}
+
+func (f normalField) Apply(b string, m Map) {
+	m[f.Key()] = b
+}
+
+type skipField struct {
+	baseField
+}
+
+func (f skipField) Apply(b string, m Map) {
+}
+
+func (f skipField) IsSaveable() bool {
+	return false
+}
+
+type namedSkipField struct {
+	baseField
+}
+
+func (f namedSkipField) Apply(b string, m Map) {
+	m[f.Key()] = b
+}
+
+func (f namedSkipField) IsSaveable() bool {
+	return false
+}
+
+type indirectField struct {
+	baseField
+}
+
+func (f indirectField) Apply(b string, m Map) {
+	v, ok := m[f.Key()]
+	if ok {
+		m[v] = b
+		return
+	}
+}
+
+type appendField struct {
+	baseField
+	previous delimiter
+}
+
+func (f appendField) Apply(b string, m Map) {
+	v, ok := m[f.Key()]
+	if ok {
+		m[f.Key()] = v + f.JoinString() + b
+		return
+	}
+	m[f.Key()] = b
+}
+
+func (f appendField) JoinString() string {
+	if f.previous == nil || f.previous.Len() == 0 {
+		return defaultJoinString
+	}
+	return f.previous.Delimiter()
+}
+
+func newField(id int, rawKey string, previous delimiter) (field, error) {
+	if len(rawKey) == 0 {
+		return newSkipField(id), nil
+	}
+
+	key, ordinal, greedy := extractKeyParts(rawKey)
+
+	// Conflicting prefix used.
+	if strings.HasPrefix(key, appendIndirectPrefix) {
+		return nil, errMixedPrefixIndirectAppend
+	}
+
+	if strings.HasPrefix(key, indirectAppendPrefix) {
+		return nil, errMixedPrefixAppendIndirect
+	}
+
+	if strings.HasPrefix(key, skipFieldPrefix) {
+		return newNamedSkipField(id, key[1:]), nil
+	}
+
+	if strings.HasPrefix(key, appendFieldPrefix) {
+		return newAppendField(id, key[1:], ordinal, greedy, previous), nil
+	}
+
+	if strings.HasPrefix(key, indirectFieldPrefix) {
+		return newIndirectField(id, key[1:]), nil
+	}
+
+	return newNormalField(id, key, ordinal, greedy), nil
+}
+
+func newSkipField(id int) skipField {
+	return skipField{baseField{id: id}}
+}
+
+func newNamedSkipField(id int, key string) namedSkipField {
+	return namedSkipField{
+		baseField{id: id, key: key},
+	}
+}
+
+func newAppendField(id int, key string, ordinal int, greedy bool, previous delimiter) appendField {
+	return appendField{
+		baseField: baseField{
+			id:      id,
+			key:     key,
+			ordinal: ordinal,
+			greedy:  greedy,
+		},
+		previous: previous,
+	}
+}
+
+func newIndirectField(id int, key string) indirectField {
+	return indirectField{
+		baseField{
+			id:  id,
+			key: key,
+		},
+	}
+}
+
+func newNormalField(id int, key string, ordinal int, greedy bool) normalField {
+	return normalField{
+		baseField{
+			id:      id,
+			key:     key,
+			ordinal: ordinal,
+			greedy:  greedy,
+		},
+	}
+}
+
+func extractKeyParts(rawKey string) (key string, ordinal int, greedy bool) {
+	m := suffixRE.FindAllStringSubmatch(rawKey, -1)
+	if len(m) == 0 {
+		return rawKey, 0, false
+	}
+
+	if m[0][3] != "" {
```

Comment:
When I saw this code my first worry is a panic in case index `3` does not exist. Same for `4` below. Should we have a check or should this never be possible based on the code path?